icefoganalytics/travel-authorization
# Add Ability for a Minimum Permission User To Copy Their Estimates to an Expenses Table

Fixes https://github.com/icefoganalytics/travel-authorization/issues/134

Relates to:
- https://github.com/icefoganalytics/travel-authorization/issues/62
- [Wireframes, Travel Authorization Workflow, 2023-10-20.pdf](https://github.com/icefoganalytics/travel-authorization/blob/157333aefdea915bb511af53625fd285fc61aad6/_Design/Wireframes%2C%20Travel%20Authorization%20Workflow%2C%202023-10-20.pdf) page 6
- [Analysis, Expense Pages, 2023-12-19](https://docs.google.com/document/d/1cWEA-0w9ro1TyajfnffkYaKL42yYnA89i9gdQysht_A/edit?pli=1#heading=h.zc2nwn7nmjlu)

# Context

1. Add ability for a minimum permission user to copy their estimates to an expenses table.
2. Copy all estimates excluding flights and meals and incidentals.
3. "Expense" tab is locked until request is approved and after travel date. The tab should be greyed out with a tooltip of why it is disabled.

**Describe the solution you'd like**
See the following screenshot. This should be very like the "Estimates" tab in the current travel authorization edit interface.
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/d4bacf89-34ab-4775-89f0-da9e7b12d2a2)

When the user has no expenses listed, it should have a "pre-fill" expenses from estimates button.
This should trigger a back-end service that:
1. copies over expenses excluding excluding flights and meals and incidentals.

## Out of scope
1. filling in meals and incidentals, these should match the estimates, but be generated independently for security reasons.
2. ability to edit estimates
3. ability to add/view reciepts
4. totals widget
5. General Ledger (G/L) coding widget.
6. Submission system.
7. Status log system
8. anything else not explicitly listed.

# Implementation

Add Expenses.UploadsController to handle expense receipt image uploading and downloading.
Fix typo in database values of Accommodations as Accomodations (with one m).
Set a default scope on the Expense model to exclude the `receiptImage` for faster load times. An alternative to this would be to make a specific table for all blob objects the way Rails does it.
Restrict setting receiptImage and fileSize independently.
Upgrade routing pattern to a less verbose form and which still keeps the requirement that routes be easily searchable.
Use `express-fileupload` as library for handling file uploads.
Add base factory concept with `attributesFor` helper method.
Use api based downloading, to avoid having to set up session cookie based access for downloads. Its not as clean or user friendly, and doesn't have the same caching benefits, but it works with what we have.

# Screenshots

Before any expenses have been created
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/f6239043-7b57-4c7a-a652-5dbc140a7695)

Once some expenses exist and some have receipts.
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/ec2f19ea-f0df-45ae-aadc-25baed523159)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Requests" page via the top drop down nav.
5. Create a new travel authorization via the "+Travel Authorization" button.
6. Check that the Expense tab is disabled, and has a tooltip explaining why.
7. Fill in the Details panel and then scroll down to the Approvals panel.
8. Click the Generate Estimate button, then save and generate.
9. Submit the travel authorization to your email, then click "Submit to Supervisor".
10. Next go to the "Manager View" from the top drop down nav.
11. You will see your recent travel request in the "Pending Approvals" top left dashboard widget.
12. Click on the travel request to load it in the manager view.
13. Scroll down to the bottom and click approve.
14. Go back to the "My Travel Requests" page via the top drop down nav.
15. Select your newly approved travel request.
16. You will now see that expenses tab is enabled.
    > Note that you are now in the "Edit" view again, and so it looks like you can edit Details and Estimates. This will need to be fixed.
17. Click on the Prefill Expenses button to prefill your expenses
    > Note that this button should lock you from doing anything, but since you are an admin through the `supervisorEmail` field it lets you create expenses. This will also need to be fixed
18. Check that you can edit expenses.
19. Check that you can delete all your expenses via the red "x" button, and prefill them again.
20. Check that you can add receipts and that once you do, the button changes to a "view receipt" button.
21. Check that you can download receipts via the view receipt button.
